File: android-client/lib/services/game_setup_service.dart
import 'dart:math';

import 'package:android_client/classes/game_data.dart';
import 'package:android_client/classes/game_room.dart';
import 'package:android_client/classes/user_game.dart';
import 'package:android_client/services/remote_games_manager_service.dart';
import 'package:android_client/services/socket_service.dart';
import 'package:android_client/services/waiting_room_service.dart';
import 'package:shared_preferences/shared_preferences.dart';

class GameSetupService {
  late String username;
  late String gameMode;
  late GameRoom gameRoom;
  final String _tokenKey = 'token';
  final SocketService _socketService = SocketService();
  late WaitingRoomService waitingRoomService = WaitingRoomService();
  final RemoteGamesManagerService _remoteGamesManagerService =
      RemoteGamesManagerService();

  late List<GameData> slides = <GameData>[];

  GameSetupService(this.gameMode) {
    getAllGames();
    getUsername();
    subscribeToGameDeletedEvent();
  }

  void subscribeToGameDeletedEvent() {
    _socketService.on('gameDeletedFromDB', (data) {
      slides.removeWhere((slide) => slide.name == data);
    });
  }

  Future<void> getAllGames() async {
    slides = await _remoteGamesManagerService.getAllRemoteGames();
  }

  Future<void> getUsername() async {
    String? tokenValue =
        (await SharedPreferences.getInstance()).getString(_tokenKey);
    if (tokenValue != null) {
      username = tokenValue;
    } else {
      username = '';
    }
  }

  void initGameRoom(bool started) {
    gameRoom = GameRoom(
      userGame: UserGame(
        gameName: null,
        creator: username,
        currentPlayers: <CurrentPlayer>[
          CurrentPlayer(username: username, isAndroid: true)
        ],
        nbDifferenceFound: 0,
        chosenDifference: -1,
        timer: 0,
        differenceFoundByPlayers: <PlayerDifferences>[
          PlayerDifferences(username: username, differencesFound: 0)
        ],
        observers: [],
      ),
      roomId: generateRandomId(),
      started: started,
      gameMode: gameMode,
      gameConstants: null,
    );
  }

  Future<bool> initGameMode(String? name) async {
    bool result = true;
    if (gameMode == "mode Temps Limité") {
      result = await initLimitedTimeMode();
    } else if (name != null) {
      result = await initClassicMode(name);
    }
    _socketService.off('gameDeletedFromDB');
    return result;
  }

  Future<bool> initClassicMode(String name) async {
    GameData? slide = getGameData(name);
    if (slide == null) {
      print("Not found");
      return false;
    }
    gameRoom.userGame.gameName = name;
    waitingRoomService.createGame(gameRoom);
    return true;
  }

  Future<bool> initLimitedTimeMode() async {
    GameData slide = randomSlide();
    if (slide == null) {
      print("Not found");
      return false;
    }
    gameRoom.userGame.gameName = slide.name;
    gameRoom.userGame.chosenDifference =
        generateRandomNumber(slide.nbDifference);
    waitingRoomService.createGame(gameRoom);
    return true;
  }

  void joinGame(String roomId, String? gameName) {
    if (gameMode == "mode Temps Limité") {
      joinLimitedTimeMode(roomId);
    } else {
      joinClassicMode(gameName, roomId);
    }
    _socketService.off('gameDeletedFromDB');
  }

  void joinClassicMode(String? gameName, String roomId) {
    GameData? slide = getGameData(gameName);
    if (slide == null) {
      print("Not found");
      return;
    }
    waitingRoomService.joinGame(username, gameMode, roomId, gameName);
  }

  void joinLimitedTimeMode(String roomId) {
    waitingRoomService.joinGame(username, gameMode, roomId, null);
  }

  GameData? getGameData(String? gameName) {
    if (gameName == null) return null;
    for (GameData game in slides) {
      if (game.name == gameName) {
        return game;
      }
    }
    return null;
  }

  void observeGame(String? gameName, String roomId) {
    if (gameMode == 'mode Classique') {
      observeClassicMode(gameName, roomId);
    } else {
      observeLimitedTimeMode(roomId);
    }
    _socketService.off('gameDeletedFromDB');
  }

  void observeClassicMode(String? gameName, String roomId) {
    GameData? slide = getGameData(gameName);
    if (slide == null) {
      print("Not found");
      return;
    }
    waitingRoomService.observeGame(username, gameMode, roomId, gameName);
  }

  void observeLimitedTimeMode(String roomId) {
    waitingRoomService.observeGame(username, gameMode, roomId, null);
  }

  String generateRandomId([int length = 16]) {
    const String characters =
        'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    final Random random = Random();
    String randomId = '';
    for (int i = 0; i < length; i++) {
      randomId += characters[random.nextInt(characters.length)];
    }
    return randomId;
  }

  GameData randomSlide() {
    return slides[Random().nextInt(slides.length)];
  }

  int generateRandomNumber(int nbOfDifferences) {
    return Random().nextInt(nbOfDifferences);
  }
}

File: windows-client/src/app/services/game-setup/game-setup.service.ts
import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { CommunicationHttpService } from '@app/services/communication-http/communication-http.service';
import { CommunicationSocketService } from '@app/services/communication-socket/communication-socket.service';
import { WaitingRoomService } from '@app/services/waiting-room/waiting-room.service';
import { GameConstants } from '@common/classes/game-constants';
import { GameData } from '@common/classes/game-data';
import { GameRoom } from '@common/classes/game-room';
import { GameModeEvents } from '@common/enums/game-mode.gateway.variables';
import { GameMode } from '@common/game-mode';
import { Subject } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class GameSetupService {
    username: string;
    gameRoom: GameRoom;
    gameConstants: GameConstants;
    gameMode: string;
    gameRoom$ = new Subject<GameRoom>();
    tokenKey = 'token';
    private slides: GameData[] = [];

    constructor(
        private readonly socketService: CommunicationSocketService,
        private communicationService: CommunicationHttpService,
        private waitingRoomService: WaitingRoomService,
        public dialog: MatDialog,
    ) {
        this.refreshGamesAfterReload();
        this.subscribeToGameDeletedEvent();
    }

    refreshGamesAfterReload() {
        this.getAllGames();
    }

    setConstants(constants: GameConstants): void {
        this.gameConstants = constants;
    }

    subscribeToGameDeletedEvent() {
        this.socketService.on(GameModeEvents.GameDeletedFromDB, (gameName: string) => {
            this.slides = this.slides.filter((game) => game.name !== gameName);
        });
    }

    getAllGames() {
        this.communicationService.getAllGames().subscribe((games) => {
            this.slides = games;
        });
    }

    getSlides(): GameData[] {
        return this.slides;
    }

    initGameRoom(started: boolean): void {
        this.username = localStorage.getItem(this.tokenKey) || '';
        this.gameRoom = {
            userGame: {
                gameName: undefined as unknown as string,
                nbDifferenceFound: 0,
                differenceFoundByPlayers: [{ username: this.username, differencesFound: 0 }],
                timer: 0,
                creator: this.username,
                chosenDifference: -1,
                currentPlayers: [
                    {
                        username: this.username,
                        isAndroid: false,
                    },
                ],
                observers: [],
            },
            roomId: this.generateRandomId(),
            started,
            gameMode: this.gameMode,
            gameConstants: undefined as unknown as GameConstants,
        };
    }

    initGameMode(gameName = undefined as unknown as string): boolean {
        let gameFound = true;
        if (gameName) {
            gameFound = this.initClassicMode(gameName);
        } else {
            gameFound = this.initLimitedTimeMode();
        }
        this.socketService.off(GameModeEvents.GameDeletedFromDB);
        return gameFound;
    }

    initClassicMode(gameName: string): boolean {
        // const slide = this.getGameData(gameName);
        // if (!slide) {
        //     this.dialog.open(MessageDialogComponent, {
        //         panelClass: 'custom-modal',
        //         data: { message: 'Jeu introuvable' },
        //     });
        //     return false;
        // }
        this.gameRoom.userGame.gameName = gameName;
        this.waitingRoomService.createGame(this.gameRoom);
        return true;
    }

    initLimitedTimeMode(): boolean {
        const gameData = this.randomSlide();
        // if (!gameData) {
        //     this.dialog.open(MessageDialogComponent, {
        //         panelClass: 'custom-modal',
        //         data: { message: 'Jeu introuvable' },
        //     });
        //     return false;
        // }
        this.gameRoom.userGame.gameName = gameData.name;
        this.gameRoom.userGame.chosenDifference = this.generateRandomNumber(gameData.nbDifference);
        this.waitingRoomService.createGame(this.gameRoom);
        return true;
    }

    joinGame(gameName = undefined as unknown as string, roomId: string): void {
        this.username = localStorage.getItem(this.tokenKey) || '';
        if (this.gameMode === GameMode.classicMode) {
            this.joinClassicMode(gameName, roomId);
        } else {
            this.joinLimitedTimeMode(roomId);
        }
        this.socketService.off(GameModeEvents.GameDeletedFromDB);
    }

    joinClassicMode(gameName: string, roomId: string): void {
        // const slide = this.getGameData(gameName);
        // if (!slide) {
        //     this.dialog.open(MessageDialogComponent, {
        //         panelClass: 'custom-modal',
        //         data: { message: 'Jeu introuvable' },
        //     });
        //     return;
        // }
        this.waitingRoomService.joinGame(this.username, this.gameMode, roomId, gameName);
    }

    joinLimitedTimeMode(roomId: string): void {
        this.waitingRoomService.joinGame(this.username, this.gameMode, roomId);
    }

    observeGame(gameName = undefined as unknown as string, roomId: string): void {
        this.username = localStorage.getItem(this.tokenKey) || '';
        if (this.gameMode === GameMode.classicMode) {
            this.observeClassicMode(gameName, roomId);
        } else {
            this.observerLimitedTimeMode(roomId);
        }
        this.socketService.off(GameModeEvents.GameDeletedFromDB);
    }

    observeClassicMode(gameName: string, roomId: string): void {
        // const slide = this.getGameData(gameName);
        // if (!slide) {
        //     this.dialog.open(MessageDialogComponent, {
        //         panelClass: 'custom-modal',
        //         data: { message: 'Jeu introuvable' },
        //     });
        //     return;
        // }
        this.waitingRoomService.observeGame(this.username, this.gameMode, roomId, gameName);
    }

    observerLimitedTimeMode(roomId: string): void {
        this.waitingRoomService.observeGame(this.username, this.gameMode, roomId);
    }

    private randomSlide(): GameData {
        return this.slides[Math.floor(Math.random() * this.slides.length)];
    }

    // private getGameData(gameName: string): GameData | undefined {
    //     return this.slides.find((game) => game.name === gameName);
    // }

    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
    private generateRandomId(length = 16): string {
        const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        let randomId = '';
        for (let i = 0; i < length; i++) {
            randomId += characters.charAt(Math.floor(Math.random() * characters.length));
        }
        return randomId;
    }

    private generateRandomNumber(nbOfdifferences: number): number {
        return Math.floor(Math.random() * nbOfdifferences);
    }
}
